### PR TITLE
configure tls cert expiration

### PR DESCRIPTION
**Pull request by pbx0**
Implementation of my last comment in this thread: https://github.com/coreos/coreos-kubernetes/issues/621

Happy to change this as that thread forms more of a consensus. I don't feel too strongly about my stance on this interim state.

cc @aaronlevy

### CODE REVIEW

**Comment by aaronlevy** [comment]
I'm starting to think we should just have:

```
--tls-ca-expiration=X
--tls-cert-expiration=y
```

Upstream kube-up uses easy-rsa which looks like it defaults to 10years for both (https://github.com/brendandburns/easy-rsa/blob/e11310d2e8dfdf01910185abc944e78053ae2bfe/easyrsa3/easyrsa#L1006-L1007)

I'm perfectly fine just defaulting to that. Or if we want we can still default to 90days/1year.

Then we just log how long we've generated a cert for in the output. e.g. "NOTE: Generated TLS assets will expire in X days". 

I don't think adding scary-named fixed duration flags is necessarily. Configurable just makes this more flexible to end user, and we can accomplish this in clearly named flags.

@philips @mike-saparov

**Comment by cgag** [comment]
cc @gtank

**Comment by aaronlevy** [comment]
Unless there are any objections, lets move forward with the option to self-select the CA/cert expiration and default to 10years (same as upstream).

If possible, I'd probably prefer that this is consumed from the config file over cli flags. This way we can have comments describing the configurable items - and can reduce flag bloat.

**Comment by pbx0** [comment]
PTAL: We now default to 10 years, warn regardless, and allow the duration to be configured in days

**Comment by aaronlevy** [comment]
We should allow for for differing expirations between CA and certs.

**Comment by pbx0** [comment]
PTAL: There are two yaml variables for the CA expiration vs cert expiration

**Comment by aaronlevy** [comment]
@pbx0 can you add the values to the config.yaml with short comment description. Also, might not hurt to update the config tests to check that we're not allowing invalid values.

/cc @colhom given it's kube-aws

**Comment by aaronlevy** [comment]
Also looks like the config unit tests are failing

File: multi-node/aws/pkg/config/config.go [hunk 1]
```diff
@@ -136,6 +136,8 @@ type Cluster struct {
 	KMSKeyARN                string            `yaml:"kmsKeyArn,omitempty"`
 	CreateRecordSet          bool              `yaml:"createRecordSet,omitempty"`
 	RecordSetTTL             int               `yaml:"recordSetTTL,omitempty"`
+	TLSCADurationDays        int               `yaml:"tlsCAValidDurationDays,omitempty"`
```
Comment (colhom) [inline multi-node/aws/pkg/config/config.go]: nit: convention is that struct field names are exactly equivalent to yaml tags if you normalize capitalization. nothing explicitly relies on this convention, but i don't see a reason to break it.

File: multi-node/aws/pkg/config/tls_config.go [hunk 2]
```diff
@@ -39,7 +40,7 @@ type CompactTLSAssets struct {
 	AdminKey      string
 }
 
-func (c *Cluster) NewTLSAssets() (*RawTLSAssets, error) {
+func (c *Cluster) NewTLSAssets(ca, cert time.Duration) (*RawTLSAssets, error) {
```
Comment (colhom) [inline multi-node/aws/pkg/config/tls_config.go]: `caDuration` and `certDuration`.

File: multi-node/aws/pkg/config/config.go (hunk 1 above)
Comment (colhom) [inline multi-node/aws/pkg/config/config.go]: @pbx0 you should set default values [here](https://github.com/pbx0/coreos-kubernetes/blob/3ff4b7e2bbc80dfdeaa50e4bfc16ec7e0a2fadb2/multi-node/aws/pkg/config/config.go#L32) for the durations. That will fix the unit tests.

**Comment by colhom** [comment]
@aaronlevy unit tests are probably failing because the default value for the config struct fields aren't set, so I imagine it's defaulting to 0.

File: multi-node/aws/pkg/config/config.go (hunk 1 above)
Comment (pbx0) [inline multi-node/aws/pkg/config/config.go]: Oh yeah, missed updating the yaml tag when I was playing with names. 

Right now I validate the values in `command_render.go` but we can add test validation as well. Also `x509.go` additionally will use default values if it receives durations <= 0.

I'll change `x509.go` do the runtime validation but not set defaults. Add a test validation step. And set the defaults the normal way instead.

File: multi-node/aws/pkg/config/tls_config.go (hunk 2 above)
Comment (pbx0) [inline multi-node/aws/pkg/config/tls_config.go]: done

**Comment by pbx0** [comment]
PTAL: Unit tests were failing because the signature of NewTLSAssets() had changed. That is no longer the case as I missed some obvious simplifications before. I think I cleaned up things a bit. 

The unit tests in `pkg/config/tls_config_test.go` should already cover this code. Runtime validation occurs in `multi-node/aws/pkg/tlsutil/x509.go`

**Comment by colhom** [comment]
LGTM

**Pull request merged**